src/structured_config.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Structured configuration for the Superset charm."""
+import logging
+from enum import Enum
+from typing import Optional
+
+from charms.data_platform_libs.v0.data_models import BaseConfigModel
+from pydantic import validator
+
+logger = logging.getLogger(__name__)
+
+
+class LogLevelType(str, Enum):
+    """Enum for the `log-level` field."""
+
+    INFO = "INFO"
+    DEBUG = "DEBUG"
+    WARNING = "WARNING"
+    ERROR = "ERROR"
+    FATAL = "FATAL"
+
+
+class StorageType(str, Enum):
+    """Enum for the `storage-type` field."""
+
+    minio = "minio"
+
+
+class CharmConfig(BaseConfigModel):
+    """Manager for the structured configuration."""
+
+    log_level: LogLevelType
+    temporal_host: str
+    storage_type: StorageType
+    storage_bucket_log: str
+    storage_bucket_state: str
+    storage_bucket_activity_payload: str
+    storage_bucket_workload_output: str
+    pod_running_ttl_minutes: int
+    pod_successful_ttl_minutes: int
+    pod_unsuccessful_ttl_minutes: int
+
+    @validator("*", pre=True)
+    @classmethod
+    def blank_string(cls, value):
+        """Check for empty strings.
+
+        Args:
+            value: configuration value
+
+        Returns:
+            None in place of empty string or value
+        """
+        if value == "":
+            return None
+        return value
+
+    @validator(
+        "pod_running_ttl_minutes", "pod_successful_ttl_minutes", "pod_unsuccessful_ttl_minutes"
+    )
+    @classmethod
+    def ttl_minutes_validator(cls, value: str) -> Optional[int]:
+        """Check validity of `*-ttl-minutes` fields.
+
+        Args:
+            value: *-ttl-minutes value
+
+        Returns:
+            int_value: integer for *-ttl-minutes configuration
+
+        Raises:
+            ValueError: in the case when the value is out of range
+        """
+        int_value = int(value)
+        if int_value > 0:
```

Comment:
i have very mixed feelings about validators in this context, cause they can potentially add a lot of code for little ROI. If people who are configuring a charm, want to put values that make absolutely no sense, maybe they shouldn't be configuring a charm :smile: . On the other hand, there could be genuine mistakes so i guess we need to find a middle ground for this... 

this is just a general comment, not strictly related to this particular validator. 
I think that the fact that we have validators for some values, might not justify adding them for all configs in the future (maybe we can focus only on those which have a significant impact).

Reply:
This is my first use of structured config and I am currently on the fence about it, as I've come to understand that the charm goes into an error state based on erroneous config, when we have usually been sending the charm into a blocked state with some message about what is wrong. This means the user must inspect the logs of `juju debug-log` to be able to understand what is wrong with the charm. I may eventually revert to this approach and perform a small amount of manual validation.